From 0d5a64f29c0e6e91bd40a4b87ac1f3781153edec Mon Sep 17 00:00:00 2001
From: Daniel Erat <derat@chromium.org>
Date: Thu, 22 Jun 2017 18:45:58 -0600
Subject: [PATCH 10/26] tlsdate: Use first proxy when multiple are supplied.

Update tlsdated's proxy-parsing code to just use the first
host it finds in a multi-proxy PAC string (e.g. "PROXY
host1.example.com:8080; PROXY host2.example.com:8080")
instead of giving up entirely.

BUG=chromium:708177
TEST=added unit tests

Change-Id: I4a0e1d88792a3ee437bad14e23a7c85e6c12bf31
Reviewed-on: https://chromium-review.googlesource.com/545181
Commit-Ready: Dan Erat <derat@chromium.org>
Tested-by: Dan Erat <derat@chromium.org>
Reviewed-by: Jorge Lucangeli Obes <jorgelo@chromium.org>
---
 src/platform-cros-util-unittest.c | 11 +++++++++--
 src/platform-cros-util.c          |  5 ++++-
 2 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/src/platform-cros-util-unittest.c b/src/platform-cros-util-unittest.c
index a78d9ec..1d23ea6 100644
--- a/src/platform-cros-util-unittest.c
+++ b/src/platform-cros-util-unittest.c
@@ -44,6 +44,12 @@ TEST(test_canonicalize_pac_parsing) {
     { "SOCKS proxy.example.com", "socks4://proxy.example.com" },
     { "SOCKS5 proxy.example.com", "socks5://proxy.example.com" },
     { "HTTPS proxy.example.com", "https://proxy.example.com" },
+    /* Multiple proxies are separated by semicolons with optional spaces. */
+    { "PROXY a.com:8080; PROXY b.com:8080", "http://a.com:8080" },
+    { "PROXY a.com:8080 ; PROXY b.com:8080", "http://a.com:8080" },
+    { "PROXY a.com; PROXY b.com", "http://a.com" },
+    { "PROXY a.com ; PROXY b.com", "http://a.com" },
+    { "PROXY a.com; DIRECT", "http://a.com" },
     /* Bad input should always result in an empty string. */
     { "", "" },
     { "BOGUS", "" },
@@ -52,12 +58,13 @@ TEST(test_canonicalize_pac_parsing) {
     { "PROXY ", "" },
     { "PROXY \n", "" },
     { "PROXY $.com", "" },
-    { "PROXY proxy.example.com:123;", "" },
-    { "PROXY proxy.example.com:123 ", "" },
     { "PROXY proxy.example.com::123", "" },
     { "PROXY proxy.example.com!", "" },
     { "PROXY http://proxy.example.com", "" },
     { "PROXY proxy_2.example.com", "" },
+    /* Don't permit empty strings at the beginning of a list. */
+    { "; PROXY b.com", "" },
+    { " ;PROXY b.com", "" },
   };
 
   char buf[256];
diff --git a/src/platform-cros-util.c b/src/platform-cros-util.c
index 112000f..69526ea 100644
--- a/src/platform-cros-util.c
+++ b/src/platform-cros-util.c
@@ -29,7 +29,10 @@ get_valid_hostport (const char *hostport, char *out, size_t len)
       return false;
     }
   *out++ = *hostport;
-  for (c = hostport + 1;  c < end && len > 0; ++c, ++out, --len)
+  /* Break on spaces and semicolons, either of which can indicate the end of the
+   * first proxy within a list. */
+  for (c = hostport + 1; c < end && len > 0 && *c != ' ' && *c != ';';
+       ++c, ++out, --len)
     {
       *out = *c;
       if (host)
-- 
2.20.1

